atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/limiter/SafeSemaphore.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.limiter;
+
+import java.io.Closeable;
+import java.util.concurrent.Semaphore;
+import java.util.function.Supplier;
+
+public final class SafeSemaphore<E extends Exception> {
+    private final Semaphore delegate;
+    private final Supplier<E> exception;
+
+    private SafeSemaphore(Semaphore delegate, Supplier<E> exception) {
+        this.delegate = delegate;
+        this.exception = exception;
+    }
+
+    public CloseablePermit acquireOrThrow() throws E {
```

Comment:
I'm not sure this suggestion is better, but I think an alternative to dealing with the exception stuff here would be:
```
public Optional<CloseablePermit> tryAcquire() { ... }
```

and then have callers do:
```
try (CloseablePermit permit = tryAcquire().orElseThrow(() -> blah)) {
    ...
}
```

Reply:
sure, it makes the limiter generic at least

Reply:
yeah bit cleaner - done / at least keeps SafeSemaphore cleaner/simpler